### ISSUE TITLE
Telegram causes music in rhythmbox (gstreamer applications) to play faster!?

### ISSUE THREAD

**Issue by cookiengineer** [telegramdesktop/tdesktop#1548]
When I startup telegram, the music in my applications play around 1.5 times the usual speed. I have not yet any kind of hint what causes the bug to appear, I just know that it _always_ happens when starting Telegram.

Hardware used: Intel NUC (i5 and Intel Audio dedicated hardware, no additions).
OS used: UbuntuGNOME 15.10

Does anybody have an idea how to debug such a thing? Is it most likely pulseaudio-related?

**Comment by cookiengineer** [comment]
I can now reproduce the bug, but I have no clue how to figure out what causes it. strace is so big that it's a couple hundred megabytes big with all the pulseaudio stuff going on.

Steps to reproduce (any Ubuntu with GNOME or GTK3 installed (also XFCE)):

- Start rhythmbox or any other gstreamer media player
- Start playing back a music track
- Open telegram binary

This will crash the pulseaudio frequency (or timer!?), so track will be played back around 1,5x times faster.

Fixes:

- When Bug occurs, go to top right corner in GNOME shell
- Open Settings from there
- Open Sound Settings

Now the bug is fixed. Can be reproduced afterwards following the steps above (telegram must be closed and re-opened, also from the systray bar).

**Comment by tdesktop-closer** [comment]
Hey there!

We're automatically closing this issue since there was no activity in this issue since 372 days ago. We therefore assume that the user has lost interest or resolved the problem on their own. Closed issues that remain inactive for a long period may get automatically locked.

Don't worry though; if this is in error, let us know with a comment and we'll be happy to reopen the issue.

Thanks!

<sub>(Please note that this is an [automated](https://github.com/auchri/github_auto_closer) comment.)</sub>

**Comment by cookiengineer** [comment]
This bug has not been magically fixed. It is still active and still telegram is causing change of frequency on devices that have not 44kHz by default.

**Comment by stek29** [comment]
Is this Telegram issue for sure?
Also, can you test other OpenAL based apps?

**Comment by cookiengineer** [comment]
I've been using the same identical setup for over 5 years now. It is telegram, for absolutely nothing-else-breaks sure. Every other application, even games, work perfectly without any issue.

Telegram keeps always resetting the frequency of the sound device, which it should not do. Eversince I reported this bug, this issue stays the same and is still causing troubles for devices that have no 44.1kHz frequency.

Yes it is confirmed, the current version of Telegram is still causing the same problems. Yes everything is as up-to-date as possible; as I'm using ArchLinux. Yes, it is still the same issue on Debian Jessie, Ubuntu 16.10, ArchLinux and even on XFCE setups.

So this is definitely a Telegram Bug, and nothing that is caused by the sound daemon. Any other application that plays back sound will be affected, even Chromium, Firefox, GNOME mplayer (and every other app you can imagine).

**Comment by gaelanlloyd** [comment]
I have the same issue too and came here to open a ticket about it.  Arch Linux here.  If I'm playing music in CMUS, VLC, Rhythmbox, etc., and then I start Telegram, the audio stream pauses for a moment, Telegram loads, and then the audio resumes at a higher pitch.  The only fix I've seen is to do `pulseaudio -k` to restart the PulseAudio daemon, but I'd rather not have to do that every time I launch Telegram.  Can this ticket be re-opened?

**Comment by mtaalas** [comment]
This is a huge issue on Windows 7 platform as well.

I use MOTU 828MKII audio interface for music production and I try to run my system on one single sampling rate no matter what. (48khz, 96khz or what ever I need or the project was made in) and every single time I start telegram it'll reset the sampling rate to 44.1khz and all the other programs are still expecting another sampling rate and audio is slowed down because of the mismatch. It's even more infuriating since I have disabled sounds on telegram and thus the audio initialization routine is completely unnecessary.

Resetting the sample-rate back to what ever it was isn't that simple since it can't be done from my audio interfaces driver unless the audio device isn't being used at all (I have to shut down _all_ applications using it first)

What ever the case may be, Telegram should respect the current sampling rate of the system and **never ever try to change it on its own**. What would be better, if I could set the sample rate that telegram requests from the system like in any decent audio software if I know what i'm doing (default to current sampling-rate by default add a tick "request sampling rate" on advanced menu).

**Comment by waaayfar3r** [comment]
Hello, is there any progress on this issue? It's been more than 3 years and the problem still persists on devices that use any sample rate other than 44.1kHz (i.e. I use an external Focusrite audio interface running at 48kHz). This is annoying and the only workaround so far is not using Telegram Desktop at all.

**Comment by Lerke** [comment]
+1 to this issue. Like @zerodayyy, I also have a Focusrite interface. Every time Telegram Desktop is launched, all of my system audio gets sped up. The only way to work around this is by having Telegram auto-launch on boot and manually restarting pulseaudio with `pulseaudio -k` and `pulseaudio -D`.

**Comment by Aryojaam** [comment]
I have the same issue. I am using Ubuntu 19.04. I have a Behringer UMC204Hd audio interface connected. I can usually fix this problem by just opening Ubuntu settings and just navigating to sound options. Does this fix work for you guys as well? @zerodayyy  @Lerke

**Comment by zerodayyy** [comment]
No, that doesn't work for me. Anyway, I'm still waiting for that to be
fixed.

**Comment by tnagler** [comment]
Same issue here with Ubuntu 19.04 and NI Komplete Audio 2 interface. For me it's "fixed" by (temporarily) switching to a different device in the Sound Settings.

**Comment by majcosta** [comment]
I have the same issue! I own a Behringer u-phoria umc202hd, and if I'm watching a person talking in a youtube video and open telegram, I get a pop, a pause, and the sound resumes but the person inhaled some helium until I `pulseaudio -k`

**Comment by Aryojaam** [comment]
I can always reproduce the bug when I do the following: 
Menu -> Calls -> Settings -> Test microphone 

I am using Ubuntu 19.04. I have a Behringer UMC204Hd

**Comment by AridaneAM** [comment]
> I can always reproduce the bug when I do the following:
> Menu -> Calls -> Settings -> Test microphone
> 
> I am using Ubuntu 19.04. I have a Behringer UMC204Hd

I have the same problem with Behringer UMC202HD and Kubuntu 19.10.

**Comment by ajaydee** [comment]
This bug is absolutely infuriating. I've stopped using telegram on desktop because of it. I've ran so many OpenAL based applications that it's obvious it's been used incorrectly here.

Please fix this bug, I'm using the desktop app to contact friend & family during isolation. My mobile internet signal is so bad, I'm having to put my phone upstairs.

I'm going to take a peek at the code. Any info on what file the openal starting code is?

**Comment by ilya-fedin** [comment]
> Ok, I see someone posted a link above to something that might be the problem.

i think he is wrong. That code converts the stream with ffmpeg, which doesn't communicate with sound system at all

More likely that the bug is caused by the [direct channels flag](https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/media/audio/media_audio.cpp#L317) or by [microphone testing](https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/media/audio/media_audio_capture.cpp#L63) which is called at start

> I've ran so many OpenAL based applications that it's obvious it's been used incorrectly here.

Would be more helpful if you provide full list of them

**Comment by ajaydee** [comment]
Blender, Dolphin, mumble, virtually every game.

Gish, Doom3, Colin McRae DiRT, ioquake3, Jedi Knight: Jedi Academy, Prey, Psychonauts, Quake 4, X3: Reunion, Alien Isolation, Minecraft.

Why am I listing software? Virtually everything uses OpenAL on Linux because it's awesome. It's quite clearly Telegram that's having the issue.

**Comment by ilya-fedin** [comment]
> Dolphin

Doesn't uses OpenAL if you're about KDE's Dolphin

**Comment by ilya-fedin** [comment]
> Minecraft

they're too AFAIK

**Comment by ilya-fedin** [comment]
> Why am I listing software? Virtually everything uses OpenAL on Linux because it's awesome.

I was hoping that you list something that is _not a game_

**Comment by ajaydee** [comment]
Blender and mumble are not games. Dolphin the emulator. Minecraft uses OpenAL, I had to enable HRTF support in the OpenAL config for it.

I've seen openal popup a good few times while install software through synaptic and looking in the depends section. I'll have a look for you. :)

**Comment by ajaydee** [comment]
I just ran a test using blender, I'm not sure if it helps. I hope it does.

I imported a 48khz wav on a speaker object. By default, Blender was set to 48Khz. Throughout this, I was playing music using my normal music app that is affected by Telegram.

During playback, I switched Blender's openal output to 44.1Khz. No Pinky & Perky. I restarted playback in blender because it stops playing audio when you switch. Nothing.

I left Blender set to 44.1Khz, and restarted it while the music was playing. No problems.

**Comment by ajaydee** [comment]
I've been looking through the blender source, and they seem to use a high-level wrapper library for OpenAL called audaspace. https://audaspace.github.io/

Maybe this might make life easier for you. :)

**Comment by john-preston** [comment]
@ajaydee If you use the version from https://desktop.telegram.org, I could try to build a version with a little bit different OpenAL init flags and you could check if it fixes the issue.

**Comment by ajaydee** [comment]
Of course, I have plenty of free time right now. I could download & compile it with the flags if you want?

Might save you bandwidth.

**Comment by ilya-fedin** [comment]
https://github.com/ilya-fedin/tdesktop/commit/7e3f5421b9f7ad63611305552631634e0ed48926
There are also a snap that I compiled for myself yesterday: https://github.com/ilya-fedin/tdesktop/actions/runs/76269832
I don't hear any distortion, however, I'm not affected to this issue

**Comment by ajaydee** [comment]
Downloading now Ilya. I don't normally use snaps, but I do have snapd installed.

**Comment by ajaydee** [comment]
Ilya, I think your version fixed it.

**Comment by john-preston** [comment]
@ajaydee Can you check the version from his action build, not snap?

Or check official snap version from snapcraft.io/telegram-desktop to see if the issue happens there?

So that we'll be sure that the change in code fixed it and not that it is the snap version that didn't have the issue even before.

**Comment by ajaydee** [comment]
It's not fixed, check my edit above.

**Comment by ilya-fedin** [comment]
So the issue (playing faster) happens while playing notifications sound/voice messages/music from message or with the call?
Calls are using libtgvoip and not openal.

**Comment by ajaydee** [comment]
The issue happens when starting telegram. It stays that way until restart of operating system.

**Comment by ilya-fedin** [comment]
Then I can't understand what do you mean by you edit :thinking:

**Comment by ajaydee** [comment]
got error message "something wrong with your soundcard" Calling me says I'm busy for friend.

Error message in terminal: "Invalid return value 0 for stream protocol"

**Comment by ilya-fedin** [comment]
This is another issue. What with this issue?

**Comment by ajaydee** [comment]
This is your snap version. Whatever you changed has uncovered this.

**Comment by ilya-fedin** [comment]
> Whatever you changed has uncovered this.

I think this is broken since this is snap and not by the change

**Comment by ilya-fedin** [comment]
The main goal is to check if this issue is present on [snap from snapstore](https://github.com/telegramdesktop/tdesktop/issues/1548#issuecomment-612804761) too and not be distracted by calls

**Comment by john-preston** [comment]
@ajaydee

1. Did snap version from Ilya fix the "Telegram causes music in rhythmbox (gstreamer applications) to play faster"?

2. Did his non-snap version fix that "causes music to play faster"?

3. Did the official snap version had this "causes music to play faster" issue?

(please check at least one of 2. or 3., no need to check both)

Calls are irrelevant to this issue, they have a completely custom code path that was not touched by the tested changes.

**Comment by ilya-fedin** [comment]
>     2\. Did his non-snap version fix that "causes music to play faster"?

There are no non-snap version on action since it was built when generic linux action was broken

**Comment by john-preston** [comment]
Ah, sorry. Then only 3. is relevant.

**Comment by ajaydee** [comment]
I'll try the official snap version soon, just chatting on phone right now. Give me an hour.

I suspect that snap won't allow access to usb devices.

**Comment by ajaydee** [comment]
Bug still exists in the official snap version. Audio goes high pitched when started.

**Comment by ilya-fedin** [comment]
I.e. bug exists in official version and fixed in version from  #action, right?

**Comment by ajaydee** [comment]
No, #action can't access soundcard because it's got the --dangerous flag (I think that's the problem).

**Comment by ilya-fedin** [comment]
And it's even not playing notification sound?

**Comment by ajaydee** [comment]
#action plays the sound for calling someone, but then hangs up with error. Didn't hear notification sounds. Not sure though.

**Comment by majcosta** [comment]
if you mean the increased sample rate to all audio after opening telegram,
I've found that killing with "pulseaudio -k" is a quick fix without
restarting the system

**Comment by ajaydee** [comment]
Yeah, I noticed an earlier post saying it. I've used it a couple of times. Just restarted to be super sure I didn't create new problems.

**Comment by ajaydee** [comment]
I was looking through the libtgvoip codebase and found this line: https://github.com/telegramdesktop/libtgvoip/blob/e422d2a80546a32ab7166a9b1058bacfc5daeefc/os/linux/AudioOutputPulse.cpp#L51

It seems to be setting up 48000 audio output for pulseaudio there.

**Comment by ajaydee** [comment]
I've fixed the issue, it wasn't your problem, it's the operating system! Anyone with a High quality soundcard needs to set their samplerate to 48000. In linux, edit **/etc/pulse/daemon.conf** and uncomment and change this line:

default-sample-rate = 44100

To:

default-sample-rate = 48000

In windows and mac, use your soundcard config to set the samplerate to 48000

I've successfully called a friend while playing music, he says the audio quality is incredible! You can be free from this problem Telegram developers. Maybe add a warning to your Download page?

**Comment by ilya-fedin** [comment]
> I was looking through the libtgvoip codebase and found this line:

I also found this line, but if the problem appears when the application starts, then it is openal, not libtgvoip

**Comment by ajaydee** [comment]
It's all cool dude, I've fixed it! Check my previous comment.

**Comment by ilya-fedin** [comment]
> It's all cool dude, I've fixed it!

I hope you try the new version without these settings, because that fix from the test snap was merged

**Comment by ajaydee** [comment]
Uh oh. Do you have a non-snap version? A deb maybe?

**Comment by ilya-fedin** [comment]
> Do you have a non-snap version?

Since linux action is fixed, I can build a test static binary

**Comment by ajaydee** [comment]
Thanks man. I'll get straight to it for you.

So you want me to set pulseaudio back to 44100 and test?

**Comment by ilya-fedin** [comment]
> So you want me to set pulseaudio back to 44100 and test?

yep

**Comment by ilya-fedin** [comment]
There are will be two artifacts when the build is done. You need the one that named as just "Telegram". Build usually takes around 2 hours.
https://github.com/ilya-fedin/tdesktop/actions/runs/77852541

**Comment by ajaydee** [comment]
Cool, I'll keep checking for it. Thank you for being awesome.

**Comment by ilya-fedin** [comment]
It's built

**Comment by ajaydee** [comment]
It won't extract, unzip says that it's part of a multipart zipfile.

**Comment by ilya-fedin** [comment]
:man_shrugging: it just opens with Engrampa

**Comment by ajaydee** [comment]
I tried with Engrampa, the file was corrupted. Maybe downloaded a bit too early. Re-downloaded, and it's extracted ok.

With pulseaudio set to 44100, it still caused the problem. Tried a call with pulseaudio set at 48000, and it works great.

I like this way of packaging telegram. Much easier than messing around with deb files IMHO.

**Comment by ilya-fedin** [comment]
> I like this way of packaging telegram.

:thinking: these binaries are at desktop.telegram.org and in github releases from the start of project AFAIK


> With pulseaudio set to 44100, it still caused the problem.

:(

**Comment by ilya-fedin** [comment]
> With pulseaudio set to 44100, it still caused the problem

from the start of app?

**Comment by ajaydee** [comment]
Yes, as soon as it logs in.

**Comment by ilya-fedin** [comment]
Could you try this artifact when it will be built?
https://github.com/ilya-fedin/tdesktop/actions/runs/78147313

**Comment by ajaydee** [comment]
Yes, no problem. :smile:

**Comment by ajaydee** [comment]
I started it with 44100, everything was ok. Music didn't go up in pitch.

Called friend, he was high pitched (very funny), but he didn't seem to think I was low or high pitched. The tone of my voice has a large range, so he probably can't tell very well.

Ended call, continued music, it had become high pitched again.

**Comment by ilya-fedin** [comment]
> I started it with 44100, everything was ok.

Then microphone check is the case :(

> Called friend, he was high pitched (very funny),

Any usage of microphone (call with libtgvoip or recording of voice messages with openal) will encounter into this issue.

Maybe yo have any examples of openal apps that uses microphone to take a look?

**Comment by ajaydee** [comment]
Blender doesn't have audio recording and mumble seems to have stopped using openAL.

The only instances of openAL for microphone recording are code snippets, as far as I can tell.

I have to go grab some food and sleep, I'll have a more thorough search for openal microphone recording software tomorrow.

**Comment by ilya-fedin** [comment]
Wonder if that can be a bug in pulseaudio and/or bug in drivers for some pro-audiocards

**Comment by ilya-fedin** [comment]
\o/
https://superuser.com/a/346085

**Comment by ajaydee** [comment]
I've just tried that, hasn't improved anything up to now. I'll leave it on. You never know, might fix something later down the line. :laughing:

**Comment by ajaydee** [comment]
How about this one: https://wiki.archlinux.org/index.php/PulseAudio/Troubleshooting#Choppy/distorted_sound

There's an openal config for output bitrate. I'll test it in a bit, my brain is still waking up.

**Comment by magnus-gross** [comment]
I actually do not think this is an Telegram or OpenAL issue at all. I also have this issue with my Focusrite Scarlett, but I can reproduce it not only with Telegram but with a simple `arecord -D 'hw:2,0' -f S32_LE --channels=2 --rate 48000 /tmp/out` (different parameters may apply for you), while Spotify plays back audio at 44100 Hz. So simply recording and playing back with a different playback rate than the recording rate will reproduce this.
I have mentioned the issue here but noone was able to help me: https://bbs.archlinux.org/viewtopic.php?id=255277

The linked PR works around the issue by not opening the recording device on start up, but I have a feeling the issue will occur again if you actually record something with Telegram.

**Comment by ilya-fedin** [comment]
> but I have a feeling the issue will occur again if you actually record something with Telegram

yep, this is mentioned in the PR

**Comment by ilya-fedin** [comment]
> I have mentioned the issue here but noone was able to help me

IMHO there are only driver bugfix can help

**Comment by magnus-gross** [comment]
> > but I have a feeling the issue will occur again if you actually record something with Telegram
> 
> yep, this is mentioned in the PR

Sry, I skipped that part. This improves things, but perhaps we should report this bug to the real project. I am not quite sure if ALSA or Pulseaudio is the culprit. However, I can certainly reproduce this issue without Telegram at all by playing something with `aplay` at 44100 Hz and recording while it is running using the above `arecord` command.

**Comment by Aokromes** [comment]
so, close this ticket?

**Comment by ilya-fedin** [comment]
> so, close this ticket?

no, imho

**Comment by magnus-gross** [comment]
Well by the looks of it it doesn't seem to be Telegram's fault at all.

I just ruled out Pulseaudio by reproducing it with direct ALSA while having no Pulseaudio running. So it is either ALSA or the Linux kernel driver for USB Audio interfaces (which all affected people seem to be using).

Perhaps one of the affected people can confirm if she/he can reproduce it with direct ALSA without Telegram like I can, so that we are sure we have the same issue?

How to reproduce:

- Find out ALSA device name: `aplay -l` => `card 2: USB [Scarlett Solo USB], Device 0: USB Audio [USB Audio]`, therefore I have `hw:2,0` (card 2, device 0).
- Playback audio: `aplay -D 'hw:2,0' /tmp/someaudio.flac` (You might need to convert the audio to the right format with `ffmpeg` and add parameters for the format in your case)
- While playback is running, record something with a different rate than the playback (in my case `aplay` was running with 8000 Hz): `arecord -D 'hw:2,0' --rate 48000 /tmp/out`
- Now your audio should sound distorted

Perhaps someone can be so kind to try to reproduce it this way without Telegram? Then we could go report the bug to the real upstream.

In case someone else has a Focusrite Scarlett, these were my exact commands:
- `aplay -D 'hw:2,0' -f S32_LE --channels=2 /tmp/somemusic.raw`
- `arecord -D 'hw:2,0' -f S32_LE --channels=2 --rate 48000 /tmp/out`

**Comment by mtaalas** [comment]
> Well by the looks of it it doesn't seem to be Telegram's fault at all.
> So it is either ALSA or the Linux kernel driver for USB Audio interfaces (which all affected people seem to be using).

Maybe I'm ignorant, but what about people having issues with Telegram Changing sampling rates when starting it on Windows systems? (Such as myself, see my previous comment https://github.com/telegramdesktop/tdesktop/issues/1548#issuecomment-406820251

This would point back to the way TG uses and initializes audio device being the issue.

Or maybe I've lost the plot here?

**Comment by magnus-gross** [comment]
You are not ignorant at all, perhaps there are multiple similar issues 
discussed in this thread, which have similar symptoms.

Telegram seems to hardcode some frequency here: 
https://github.com/telegramdesktop/tdesktop/blob/1e5aa2a78d0a0ba06d95c63e746d7b795fe838b8/Telegram/SourceFiles/media/audio/media_audio.h#L52
Which is then used here: 
https://github.com/telegramdesktop/tdesktop/blob/0c0fc46b90d318d6f8a258537abe9bbd11732239/Telegram/SourceFiles/media/audio/media_audio_capture.cpp#L176

Perhaps something like the snippet here should be used: 
https://github.com/kcat/openal-soft/issues/350#issuecomment-543453776

But I don't think this would cause a problem, as Pulseaudio would 
resample the audio stream as needed (and I am sure Windows does the 
same). Also the people with the sample rate problem report Telegram 
setting the sample rate to 44100 Hz, despite it being hardcoded to 48k.


I think there are definetely at least two different bugs mentioned in 
this thread.

**Comment by ilya-fedin** [comment]
> but what about people having issues with Telegram Changing sampling rates when starting it on Windows systems?

Windows can have buggy drivers too
If this isn't a driver issue, then all users should be affected, right? But HDA codecs doesn't have this issue not at one of the systems.
btw, you are the only one who have this issue on Windows in this ticket :thinking:

**Comment by magnus-gross** [comment]
Yes, when I switch to my HDA card I dont have this problem either. Together with the fact that almost all affected people use an USB audio interface, this sure sounds like a bug in the Linux Alsa USB driver.

**Comment by ilya-fedin** [comment]
I think this issue is not so common because most programs uses 44100 Hz by default
Also, I wonder, will this issue be if set capture frequency lower than 44100 Hz?

**Comment by Lerke** [comment]
@magnus-gross Thank you for your clearly deep insights in this problem. 

> Perhaps one of the affected people can confirm if she/he can reproduce it with direct ALSA without Telegram like I can, so that we are sure we have the same issue?

I have tried the exact same steps you have, as I have a similar interface (Focusrite 2i2). My sound gets distorted as well after running the commands you suggested. For what it's worth: The distortion is 'actual' distortion though, as in: crackling audio and a pitch-shift down, in my case. Instead of the rest of the system pitch-shifting up. If I have just Spotify open, and only run `arecord -D...` from the command line, I get the exact same problem as if I were to open Telegram!

Thank you, and all other contributors, for your collective efforts.

**Comment by magnus-gross** [comment]
@Lerke Thanks for confirming, so this is indeed an Alsa problem. I have submitted a kernel bug report now: https://bugzilla.kernel.org/show_bug.cgi?id=207617

Hopefully this can be fixed there.

**Comment by mtaalas** [comment]
@magnus-gross 
> Telegram seems to hardcode some frequency here...

And that's the issue I think. My MOTU828 MK2 (and almost all professional audio interfaces I've used thus far) are designed to work with one single samplerate at a time and can be externally clocked to one rate if necessary and no resampling happening on the driver level. If you force it to be something else suddenly, it'll change the whole interface to that sampling rate and every program that assumes a different one (a DAW like REAPER for example that's being run, you always set a particular samplerate for your current project when you run it and it has direct access to the hardware through ASIO drivers given by MOTU) will go haywire, or Chrome that still thinks they're on different sampling rate until it's restarted.

This is evident by the fact that sampling rate cannot be changed with the audio interfaces own setup tool unless you close all other software that's using it first.

I have yet tried to clock it externally since those master clocks aren't cheap. 

Software shouldn't assume a driver does any resampling, it should ask what sampling rate the system's at and have it's own internal resampling to that target. And yes, to update my driver by MOTU, I'd need to dish out another 500€+ or so for newer MK3 unit and even then it might not be "fixed"... it's dangerous to assume resampling on driver level in my opinion.

Sorry for being so direct and harsh... I'm just of different opinion on who's problem it should be... maybe the wrong thread for this discussion? Maybe we need to start a new one?

Anyway... I hope this will get solved somehow :)

**Comment by magnus-gross** [comment]
@mtaalas Yes you are right, that is a problem as well separate from the Alsa driver bug. I didn't notice this before, because my audio interface is at 48k Hz anyway, but I just set it to 44100 Hz and Telegram switched it back to 48k Hz.

So at last there is indeed something wrong in Telegram. I will maybe try looking into it and fixing it (though this is not my priority, as in my case Pulseaudio can handle resampling just fine - but I agree with you that this is not ideal)

**Comment by tsundokul** [comment]
A fix that worked for me (`SMP Debian 5.6.14-1kali1`) was setting the sample rate to 48000 in `/etc/pulse/daemon.conf` and restarting the daemon
```
default-sample-rate = 48000
```

**Comment by clbustos** [comment]
The fix by tsundokul (fix sample rate to 48000 on pulseaudio) works for me on a Audient ID14

**Comment by magnus-gross** [comment]
Interesting, for me the workaround is to use `default-sample-rate=44100` 
in Pulseaudio. Using `48000` reproduces the problem for me.

What is your audio interface, just out of interest? I still think there 
are multiple different bugs being talked about here.

**Comment by clbustos** [comment]
@magnus-gross It is listed as "Audient Audient iD14 at usb-0000:00:14.0-1, high speed", with sample spec by default of "s32le 10ch 44100Hz"

**Comment by PandaTheSlayer** [comment]
Any hacks to fix it on windows?

**Comment by ilya-fedin** [comment]
> Any hacks to fix it on windows?

I guess the only way to fix it on Windows is to inform driver makers about this issue

**Comment by mtaalas** [comment]
> > Any hacks to fix it on windows?
> 
> I guess the only way to fix it on Windows is to inform driver makers about this issue

Ummm... I have this suspicion that relying on the audio interface driver/hardware for resampling isn't the smartest idea. In many cases professional audio interfaces are designed such that they definitely don't use resampling by design and are designed to work with one sample rate that's defined by the user and never change it by requesting something else.

Many interfaces are externally clocked to a separate master clock that has no clue about applications on the computer requesting a changes in sample rate.

Thus the application should ask the system for what sample rate it's running at and use internal resampling to respect what it expects.

This is just my opinion based on over a decade as an audio technician. :)

**Comment by PandaTheSlayer** [comment]
Ok, I switched sample rate to 96khz in focusrite settings and this fixed the problem, but I don't think that it's a good way to solve a problem

**Comment by ilya-fedin** [comment]
> Thus the application should ask the system for what sample rate it's running at and use internal resampling to respect what it expects.

openal documentation says another thing: an app shouldn't request sample rate from the system, it should just request any sample rate that it wants instead and the system must resample it as needed.

**Comment by magnus-gross** [comment]
This is not entirely true. Yes, your audio interface does not do 
resampling, but there can still happen resampling inside the audio stack 
of your operating system.

For example on Linux both Pulseaudio and alsalib provide mechanisms to 
resample.
Of course it goes without saying that these libraries try to avoid 
resampling if possible, i.e. if Telegram is the only stream playing and 
it wants to play at 48000 Hz, but your audio interface is currently 
configured with 44100 Hz but is known to also support 48000 Hz, then it 
will configure the device to use 48000 Hz instead of doing an expensive 
and unnecessary resample.

That being said, these libraries have options to always resample and in 
the case of Jack2, this is the default as you configure the sample rate 
on server startup.

In any case, these kinds of problems should not appear and in fact they 
don't seem to appear on HDA devices as all reported cases use some kind 
of USB audio interface.

**Comment by ilya-fedin** [comment]
Is something changed with 2.4?

**Comment by ilya-fedin** [comment]
Anyone?

**Comment by magnus-gross** [comment]
Hi, no nothing is changed in 2.4.3. Starting Telegram will instantly 
produce distorted audio unless the workaround with the default 
Pulseaudio sample rate is used.

**Comment by ilya-fedin** [comment]
I'm more interested to hear feedback from @mtaalas or @PandaTheSlayer

**Comment by PandaTheSlayer** [comment]
@ilya-fedin problem still exist on 44.1khz and 48khz. It disappears only at 96khz and over

**Comment by ilya-fedin** [comment]
>  problem still exist on 44.1khz and 48khz. It disappears only at 96khz and over

with 2.4.3?

**Comment by PandaTheSlayer** [comment]
> > problem still exist on 44.1khz and 48khz. It disappears only at 96khz and over
> 
> with 2.4.3?

![image](https://user-images.githubusercontent.com/15297141/95727153-7d283180-0c82-11eb-8a01-a189253d41ea.png)
Yep

**Comment by ilya-fedin** [comment]
That's weird, tdesktop uses default frequency for playback now
https://github.com/telegramdesktop/tdesktop/commit/d86b4659d646eb8c2ed4c86c1ede93895de6c8c4

**Comment by magnus-gross** [comment]
Playback isn't really the problem anyway, it is only recording that 
creates the problem. Unfortunately enumerating the capture devices is 
enough to reproduce: https://github.com/kcat/openal-soft/issues/457

I looked into this a while ago and also tried not hardcoding the default 
frequency, but unfortunately the problem isn't as easy as that.

The problem gets weirder and weirder the more you look into it.

In case anyone wants to continue where I left off: I remember the 
problem starting to appear as soon as alsa calls 
_snd_pcm_hw_params_internal 
https://github.com/alsa-project/alsa-lib/blob/4f90392f07e8822d1984ed990f622ad36022a4a3/src/pcm/pcm_params.c#L2383

I think it is also quite interesting that the problem appears on both 
Windows and Linux so perhaps both audio stacks do something wrong.

**Comment by ilya-fedin** [comment]
Or maybe there something wrong with openal's capture capability... That explains why tdesktop is the only affected application that uses openal then, since all others are only playing audio :(

**Comment by Pixelnarium** [comment]
I know it would not be a fix at all, but maybe an option to disable recording would help people like me. I do not use Telegram for calls.

**Comment by Pixelnarium** [comment]
One thing I saw after a quick look is that OpenAL asks for a specific sample rate from pulseaudio https://github.com/kcat/openal-soft/blob/master/alc/backends/pulseaudio.cpp#L620

Maybe this confuses ALSA or the hardware in this case. But I don't know pulseaudio that much.

**Comment by magnus-gross** [comment]
Good find. I looked into it but unfortunately it was a dead end again.

I tried changing the hardcoded frequency and sample format to my native 
audio interface format, but the distorted audio still occurs.

The problem appears as soon as OpenAL calls into pa_stream_new.

I also tried using pa_stream_new_extended instead (see 
https://freedesktop.org/software/pulseaudio/doxygen/stream_8h.html#acf41f201c5e6ac94b7fcbc8626b6c933), 
where instead of hardcoding a sample rate and format, you can give a 
list of acceptable formats and Pulseaudio will figure out what is the 
most suitable format. Sadly this was a dead end too.

I think at the end of the day this is still an Alsa problem, openal-soft 
is not really doing anything wrong even when that hardcoded probe 
frequency seems a little suspicious.

**Comment by ilya-fedin** [comment]
> I also tried using pa_stream_new_extended instead

Why just not to remove `PA_STREAM_FIX_FORMAT | PA_STREAM_FIX_RATE | PA_STREAM_FIX_CHANNEL` flags?

**Comment by magnus-gross** [comment]
You are right that would have the same effect.

I just tested it though and it does not work either.

**Comment by flamingspaz** [comment]
Same issue with a Focusrite Scarlett Solo gen1, setting pulseaudio to 96khz doesn't fix the issue. fwiw the Zoom linux client does the same thing.

**Comment by ilya-fedin** [comment]
Wow, it is good to know that not only tdesktop is affected

**Comment by Aokromes** [comment]
sooo, 3rd party?

**Comment by ilya-fedin** [comment]
> sooo, 3rd party?

We still don't know why exactly this happens :(

**Comment by Pixelnarium** [comment]
Some observations on my (!) system (sorry if this was already known, this issue is kinda long). The interface is a Steinberg UR22 MK2.

- in my case the sound gets slower
- this seems to happen only when the default input source is set to the interface. Using the build-in microphone port on the mainboard as default does not trigger the issue. Changing it to the interface after starting Telegram seems to work just fine. Verified with multiple tries and reboots.
- the sampling rates Pulseaudio reports via `pacmd list-sinks` or pacmd list-sources` do not change while triggering the issue
- just changing the different device profiles in pavucontrol (analog input + digital input, digital input + analog output, ...) can trigger is issue
- having different sampling rates set may or may not trigger the issue, depending on the current sampling rate. i.e. for me it does not trigger when everything is set to 44.1khz but does trigger when the output is set to 48khz and the input is set to 44.1khz.
- at least my interface was pretty troublesome (stuttering, audio dropouts, ...) on Windows 10 after the 1803 update, I think a later Yamaha driver update fixed some (?) issues
- you can trigger the issue with multiple applications (i.e. pavucontrol, OBS studio, pure OpenAL, ALSA helper tools)
- you can also trigger the issue by just opening the GNOME audio settings while the "wrong sample combination" and the interface input source is currently active
- it does not trigger when just opening and using the GNOME Sound Recorder
- GNOME and all its daemons have nothing to do with it, it also happens on i3wm

What I did **not** test:
- setting up a pure ALSA system. When I bought the interface a couple of years ago I had some troubles using it on a pure ALSA system. The ALSA subsystem crashed when I started Audacity.
- setting up a pure JACK system (I never got JACK running nicely on any system)
- setting up a PipeWire system (too new as an audio server, barely works)

Personally I think all this hardware needs some kind of quirks settings at the kernel level if it is not a kernel bug for USB Audio in general. Since nothing reports any issues or changes while playing I think the hardware gets confused when the output sampling rate is not in a good combination with the input sampling rate when the specific hardware call happens.

Since I could not find the Kernel Bugzilla entry in here, here is the link from the OpenAL issue above:

https://bugzilla.kernel.org/show_bug.cgi?id=207617

**Comment by Pixelnarium** [comment]
Okay, since this issue is interesting I tried something more. I set up a pure ALSA system and tried some things:

- at least for me this issue does **not** happen on a pure ALSA system (i.e. Archlinux base install)
  - `aplay` and `arecord` are working nicely in parallel no matter what sample rate is used
  - the OpenAL sample code also does not trigger the issue when aplay plays the music
- the issue seems to be triggered or caused by Pulseaudio
  - installing & starting pulseaudio and playing a 48khz music file via `mpv` and running the OpenAL sample code **does** trigger the issue
  - installing pulseaudio-alsa sets up the alsa wrapper so that ALSA calls get redirected to Pulseaudio
    - running `aplay` and `arecord` after installing said package the issue also happens

I found a similar issue on the Pulseaudio Bug Tracker: https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/issues/944

Maybe we should ask there.

**Comment by magnus-gross** [comment]
Yup I have the same pulseaudio-alsa package installed. I did not know that it redirects Alsa sound to pulseaudio, but if it does like you said it does, then this is indeed a pulseaudio problem, which is good news because that would be way easier to debug than a bug deep inside the kernel audio system.

**Comment by ilya-fedin** [comment]
Today I noticed that pipewire finally got sound support and could be enabled on arch by installing pipewire-alsa, pipewire-pulse, pipewire-pulse-dropin and lib32-pipewire-pulse-dropin package (the last one is needed if 32-bit applications like wine are used, two last ones are from AUR) and editing the last line of /etc/pipewire/pipewire.conf to `exec /usr/bin/pipewire-media-session -e bluez5,pulse-bridge`

_In case if anyone want to test if this bug is present with pipewire and report it to them, since pulseaudio development seems to be very slow_

**Comment by avemilia** [comment]
I also have audio speed up on 2.5.5 beta, but only when I'm calling or receiving a call. Just launching telegram and receiving sound notifications is fine. before I accept a call I physically turn on +48V mic on usb sound card. the fix after the call is to kill pulseaudio to make it restart.

**Comment by ilya-fedin** [comment]
There's probably nothing we can do about that since there's a bug with handling different sample rates on pulseaudio side

**Comment by OlliC** [comment]
I am also using a Steinberg UR22 MK2 interface and had the same issue. I reinstalled my PC and forgot that i had set a fixed sample rate of 48kHz in my /etc/pulse/daemon.conf before, so the issue popped up again. With a fixed sample rate the issue does not happen. This is a general issue with external USB audio interfaces and pulseaudio.

These are the settings i use:
```
resample-method = speex-float-10
default-sample-rate = 48000
alternate-sample-rate = 48000
```

**Comment by avemilia** [comment]
This is not the case for me -- people I talk to still sound like they are on helium baloons. With the following config
```
❯ grep -ve '^;' /etc/pulse/daemon.conf | grep -ve '^#' | grep '\S'
resample-method = speex-float-10
flat-volumes = no
default-sample-rate = 48000
alternate-sample-rate = 48000
```

and Audient iD4 as sound card, in pulseeffects I see:
![2021-01-31-103712_1796x204_scrot](https://user-images.githubusercontent.com/46303212/106380238-1a81af80-63b1-11eb-971b-4d7a000e3576.png)
![2021-01-31-103756_860x77_scrot](https://user-images.githubusercontent.com/46303212/106380240-1d7ca000-63b1-11eb-9cf5-c6cf6fbce756.png)

Does this mean PA ignores my settings? Weird.

**Comment by ilya-fedin** [comment]
> Does this mean PA ignores my settings? Weird.

Maybe some another application is forcing PA to 44100? I've seen that PA switches to another sample rate when some application requests it and there are no other application playing at the same time and it doesn't reset to default sample rate when the playback is finished

**Comment by vimpostor** [comment]
I just set up Pipewire to replace Pulseaudio on my system. I can report that the problem does no longer appear when using Pipewire on my system.

**Comment by avemilia** [comment]
Same. I have installed pipewire 0.3.19 and the problem is gone.
